src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfAnalyzer.cs

```diff
@@ -0,0 +1,231 @@
+                        ParameterInfo.GetParameterInfo(stringType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+                var charAndComparisonTypeArgumentContainsMethod = stringContainsMethods.GetFirstOrDefaultMemberWithParameterInfos(
+                        ParameterInfo.GetParameterInfo(charType),
+                        ParameterInfo.GetParameterInfo(stringComparisonType));
+
+                // Roslyn doesn't yet support "FindAllReferences" at a file/block level. So instead, find references to local variables in this block.
+                context.RegisterOperationBlockStartAction(context =>
+                {
+                    if (!(context.OwningSymbol is IMethodSymbol method))
+                    {
+                        return;
+                    }
+
+                    List<Location>? leftOperandInvocationLocations = null;
+                    PooledConcurrentDictionary<string, int> variableNameToNumberOfReferences = PooledConcurrentDictionary<string, int>.GetInstance();
+                    PooledConcurrentDictionary<string, ImmutableArray<Location>> variableNameToLocationsMap = PooledConcurrentDictionary<string, ImmutableArray<Location>>.GetInstance();
+
+                    context.RegisterOperationAction(context =>
+                    {
+                        ILocalReferenceOperation localReference = (ILocalReferenceOperation)context.Operation;
+                        ILocalSymbol symbol = localReference.Local;
+                        variableNameToNumberOfReferences.AddOrUpdate(symbol.Name, 1, (name, addValue) =>
+                        {
+                            return addValue + 1;
+                        });
+                    },
+                    OperationKind.LocalReference);
+
+                    context.RegisterOperationAction(HandleStringIndexOfMethod, OperationKind.Binary);
+
+                    context.RegisterOperationBlockEndAction(CheckInvocationOrNumberOfReferencesAndReportDiagnostic);
+
+                    void HandleStringIndexOfMethod(OperationAnalysisContext context)
+                    {
+                        IBinaryOperation blockOperation = (IBinaryOperation)context.Operation;
+                        var operatorKind = blockOperation.OperatorKind;
+                        if (operatorKind != BinaryOperatorKind.Equals && operatorKind != BinaryOperatorKind.GreaterThanOrEqual)
+                        {
+                            return;
+                        }
+
+                        // Check that the right hand side is a -1(IUnary Operation) or 0(ILiteralOperation)
+                        var rightOperand = blockOperation.RightOperand;
+                        if ((rightOperand is IUnaryOperation unaryOperation || rightOperand is ILiteralOperation literalOperation) && rightOperand.ConstantValue.HasValue && rightOperand.ConstantValue.Value is int intValue && (intValue == -1 || intValue == 0))
```

Comment:
Adding so many nesting if statements makes the code harder to read. Instead invert the if condition and bail out when false to make the code more easier to read.